Simulation updates

AdarshKaran opened the pull request:
Summary of changes addressing Issue #48 

**rox_description**
- Updated `gazebo.xacro` for migration to Modern Gazebo (Ionic).
- Added modular URDF xacros for `ur_arm` and `elite_arm`.
- Refactored `rox.urdf.xacro` to be modular.
- Resolved movement issues in Argo drive in Gazebo.
- Adjusted positions of various robot components in the URDF.
- Added and fine-tuned dynamic properties for wheels and joints.
- Fixed stability issues with the robot when the arm is mounted.
- Thoroughly tested `/odom` and `/cmd_vel` topics.
- Documented updates in `CHANGELOG`.

**rox_bringup**
- Added ROS 2 Control and Joint Trajectory Controller configurations for UR and Elite arms.
- Updated `gz_bridge_config.yaml` to support Modern Gazebo migration and updated `/cmd_vel` message type.
- Modified `bringup_sim_launch.py` to accommodate the new URDF format and include necessary ROS 2 Control nodes for arms.
- Updated teleop publish message type for `/cmd_vel` topic.
- Added model paths to `GZ_SIM_RESOURCE_PATH` environment variable in the launch file.
- Set `use_ac/dc` parameter in the simulation launch file and updated arm position (height) on the cabinet for the AC variant.
- Adapted UR arm URDF in `ur_description` package for Rolling distro changes.
- Conditionally added `elite_description` to the environment variable if an Elite arm is in use.

**General**
- Updated dependencies.
- Restructured packages.

## **NOTE!!**
For simulation, set:
```yaml
speed_scaling_interface_name: ""
```
in the respective `ur_controllers.yaml` file

AdarshKaran committed:
Migrate to Modern Gazebo & Modularize URDF for UR & Elite Arms

AdarshKaran committed:
Added UR and Elite xacros and controller files

AdarshKaran committed:
Updated README.md and added contributor

AdarshKaran committed:
Fixed issues with rox argo, diff and trike drive in simulation

AdarshKaran committed:
minor updates to launch files and urdf

AdarshKaran committed:
Resolving conflicts and merging

AdarshKaran committed:
Updated Readme, speed_scaling_interface_name was set to work with real robot

padhupradheep commented on rox_bringup/configs/ur/ur10e/ur_controllers.yaml:
```diff
@@ -107,10 +107,10 @@ scaled_joint_trajectory_controller:
 
     # Use this for simulation
-    speed_scaling_interface_name: ""
+    # speed_scaling_interface_name: ""
```
does speed scaling always required for simuiation?

AdarshKaran replied:
Yes. In the [scaled_joint_trajectory_controller](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/blob/main/ur_controllers/src/scaled_joint_trajectory_controller.cpp) code, there is a condition (lines 53–58) that checks whether the parameter speed_scaling_interface_name is empty
![Screenshot from 2025-03-05 19-20-03](https://github.com/user-attachments/assets/dd1da51f-6df9-4516-a2e6-c58b6d2f11ee)

According to this code, an empty value for speed_scaling_interface_name is expected if no scaling interface is available. I have tested both the cases:
<br>
**With an empty value in the YAML**
`speed_scaling_interface_name: ""`
Result:

![Screenshot from 2025-03-05 19-15-53](https://github.com/user-attachments/assets/dda1ec5e-d020-4358-b2d8-1e3beb05abfc)
<br>
**When completely removing the parameter from the YAML**
Result:
![Screenshot from 2025-03-05 19-16-26](https://github.com/user-attachments/assets/2561213c-cf19-49a2-a6ab-3f47b5748e93)
<br>
In GZ sim no scaling interface is available [GitHub](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/pull/1172#issue-2621734954)

the code that i gave before points to this change mentioned in the [comment](https://github.com/UniversalRobots/Universal_Robots_ROS2_GZ_Simulation/issues/76#issuecomment-2545892116
)
![Screenshot from 2025-03-05 19-24-25](https://github.com/user-attachments/assets/daf49bd8-5311-4202-9b1d-f53312a57ba5)

padhupradheep commented on README.md:
```diff
@@ -1 +1,79 @@
 # **rox**
+
```
Readme should point always to the docs..

AdarshKaran committed:
Refactored simulation controller setup,removed 'speed_scaling_interface_name' and 'scaled_joint_trajectory_controller', added param_file_utils.py for dynamic prefix substitution,updated CMakeLists.txt

AdarshKaran committed:
Update GIF positions in README

AdarshKaran commented:
1. Removed `speed_scaling_interface_name` for simulation from controller YAML and replaced `scaled_joint_trajectory_controller` with `joint_trajectory_controller` as the default in `bringup_sim_launch.py`.
2. Added `param_file_utils.py`: A helper function for dynamic prefix substitution in simulation_controllers, modifying the launch file accordingly.
3. Simplified controller configuration by merging individual arm controllers into a single `controllers.yaml` per arm manufacturer and updated `CMakeLists.txt` to install Python modules and include the new utility file.
4. Refactored the launch file with new conditions to ensure edge cases are covered

AdarshKaran committed:
added use_mock_hardware and mock_sensor_commands arg in  rox.urdf.xacro and ur_arm.urdf.xacro; added  arg in , modified the launch file to make it work with mock hardware and ur arm

AdarshKaran commented:
`rox_description`

- added `use_mock_hardware` and `mock_sensor_commands` arg in  `rox.urdf.xacro` and `ur_arm.urdf.xacro`

`rox_bringup`

- added `use_mock_hardware` arg in `bringup_launch.py`
- refactored the launch file to make it work with mock hardware and ur arm
- tested the changes

Commands
---
using mock hardware; no simulation

- `ros2 launch rox_bringup bringup_launch.py arm_type:=ur10 robot_ip:=yyy.yyy.yyy.yyy use_mock_arm:=True initial_joint_controller:=joint_trajectory_controller`
- `ros2 launch neo_rox_moveit2 neo_ur_moveit.launch.py arm_type:=ur10 prefix:=ur10 launch_rviz:=True use_mock_hardware:=True`

using gz sim

- `ros2 launch rox_bringup bringup_sim_launch.py arm_type:=ur10`
- `ros2 launch neo_rox_moveit2 neo_ur_moveit.launch.py arm_type:=ur10  prefix:=ur10 launch_rviz:=True use_gz:=True`

AdarshKaran committed:
Added use_rviz launch argument, changed the default value of nav2_params_file to empty and added condition to set it based on rox_type

AdarshKaran commented:
- Added the missing use_rviz launch argument to rox_navigation navigation.launch.py
- Changed the default value of nav2_params_file to empty value. This will make it flexible for the user to enter a custom params file if needed. A new condition to check if this file is empty has been added. If the user chooses not give give any custom params file, it is set to the default one based on rox_type.

padhupradheep committed:
updating bringup launch

AdarshKaran committed:
Refactored launch files, fixed urdf indentations, changed outdated 'joint_properties' tag in urdf to 'dynamics', added initial_joint_controller condition in bringup_launch, removed meca drive and frame type

AdarshKaran commented:
- Merged with changes done by Pradheep
- Refactored launch files
- Fixed urdf indentations and changed outdated `joint_properties` tag in urdf to `dynamics`
- Added `initial_joint_controller` condition in `bringup_launch.py` as this is required to ensure the mock hardware works correctly
- Removed `meca` drive from both real robot and simulation launch files and `frame type`, `frame_len` arg from all urdf (ensuring only `short` frame is supported)

TO DO
- [ ] Namespace and multi-robot support

padhupradheep commented on rox_bringup/launch/bringup_sim_launch.py:
```diff
@@ -196,34 +181,34 @@ def generate_launch_description():
+            description='Arm Types\n\t'        
         )
-
+    
```
remove

padhupradheep commented on rox_description/urdf/xacros/argo_drive.xacro:
```diff
@@ -8,25 +8,25 @@
     <xacro:argo_caster name="caster_back_right" parent="${parent}"
         vis_rot="0 0 ${pi}" vis_pos="0 0 0"
         joint_type="${joint_typ}">
-        <origin rpy="0 0 0" xyz="-${0.239 + frame_len} -0.229 0.123"/>
```
could you remove `${}`

padhupradheep commented on rox_description/urdf/xacros/diff_drive.xacro:
```diff
@@ -8,28 +8,28 @@
         vis_rot="-${pi/2} 0 ${pi/2}" vis_pos="0 0 0" collision_pos="0.045 0 0"
         joint_type="fixed">
-        <origin rpy="0 0 0" xyz="${0.2045 + frame_len + 0.045} -0.220 0.124"/>
+        <origin rpy="0 0 0" xyz="${0.2045 + 0.045} -0.220 0.124"/>
```
now we don't have the frame length, can you give the summed up value? `${0.2045 + 0.045}`

padhupradheep commented on rox_description/urdf/xacros/diff_drive.xacro:
```diff
@@ -8,28 +8,28 @@
         vis_rot="-${pi/2} 0 ${pi/2}" vis_pos="0 0 0" collision_pos="0.045 0 0"
         joint_type="fixed">
-        <origin rpy="0 0 0" xyz="${0.2045 + frame_len + 0.045} 0.220 0.124"/>
+        <origin rpy="0 0 0" xyz="${0.2045 + 0.045} 0.220 0.124"/>
```
same as above

padhupradheep commented on rox_description/urdf/xacros/diff_drive.xacro:
```diff
@@ -8,28 +8,28 @@
         vis_rot="-${pi/2} 0 ${pi/2}" vis_pos="0 0 0" collision_pos="0.045 0 0"
         joint_type="fixed">
-        <origin rpy="0 0 0" xyz="-${0.2945 + frame_len - 0.045} -0.220 0.124"/>
+        <origin rpy="0 0 0" xyz="-${0.2945 - 0.045} -0.220 0.124"/>
```
same

padhupradheep commented on rox_description/urdf/xacros/diff_drive.xacro:
```diff
@@ -8,28 +8,28 @@
         vis_rot="-${pi/2} 0 ${pi/2}" vis_pos="0 0 0" collision_pos="0.045 0 0"
         joint_type="fixed">
-        <origin rpy="0 0 0" xyz="-${0.2945 + frame_len - 0.045} 0.220 0.124"/>
+        <origin rpy="0 0 0" xyz="-${0.2945 - 0.045} 0.220 0.124"/>
```
same

padhupradheep commented on rox_description/urdf/xacros/trike_drive.xacro:
```diff
@@ -8,7 +8,7 @@
         vis_rot="-${pi/2} 0 -${pi/2}" vis_pos="0 0 0" collision_pos="-0.045 0 0"
         joint_type="fixed">
-        <origin rpy="0 0 0" xyz="-${0.2785 - 0.045 + frame_len} 0.0 0.124"/>
+        <origin rpy="0 0 0" xyz="-${0.2785 - 0.045} 0.0 0.124"/>
```
same

padhupradheep commented on rox_navigation/launch/navigation.launch.py:
```diff
@@ -164,9 +180,11 @@ def generate_launch_description():
+            'nav2_params_file', default_value=os.path.join(
+                get_package_share_directory('rox_navigation'),
+                'configs',
+                'navigation_omni.yaml'),
```
would be cool if we are able to select the diff or trike option from the launch?

padhupradheep reviewed (changes requested):
Almost done! Some minors!

AdarshKaran committed:
Fixed indentations and formatting issues

padhupradheep commented:
That was one big PR. Now our ROX simulation is going to be live.. thanks for all the work Adarsh.

I will update the docs. If I don't do it within this week, could you please nudge me once on that.  
If I don't also do the cherrypick for jazzy let me know..

AdarshKaran commented:
> That was one big PR. Now our ROX simulation is going to be live.. thanks for all the work Adarsh.
> 
> I will update the docs. If I don't do it within this week, could you please nudge me once on that. If I don't also do the cherrypick for jazzy let me know..

Thank You Pradheep!
Sure :+1:
